File: Zig/p002.zig
const std = @import("std");
const print = std.debug.print;

pub fn main() void {
    var f: Fib = Fib.init(1, 2);

    const s = blk: {
        var c: u32 = 0;
        while (f.curr <= 4000000) {
            if (f.curr % 2 == 0) {
                c += f.curr;
            }
            f = f.next_fib();
        }
        break :blk c;
    };
    print("result: {}\n", .{s});
}

const Fib = struct {
    curr: u32,
    next: u32,

    pub fn init(x1: u32, x2: u32) Fib {
        return Fib{
            .curr = x1,
            .next = x2,
        };
    }

    fn next_fib(self: Fib) Fib {
        return Fib{
            .curr = self.next,
            .next = self.curr + self.next,
        };
    }
};

File: Julia/p002.jl
using BenchmarkTools

mutable struct Fib{T}
    curr::T
    next::T
end

function next(f::Fib{T}) where T <: Number
    (f.curr, f.next) = (f.next, f.next + f.curr)
end

function p002()
    init = Fib(1, 1)
    s = 0
    while init.next <= 400_0000
        next(init)
        s += iseven(init.curr) ?
            init.curr :
            0
    end
    return s
end

@btime p002()

